cli/src/lib.rs

```diff
@@ -0,0 +1,74 @@
+use std::{ffi::OsString, fs, path::PathBuf};
+
+use benchmark_common::{parse_raw, IncomingMessage, OutgoingMessage, SuiteStatus};
+use tty::{SerialConnection, SuiteConnection};
+
+pub mod tty;
+
+/// Benchmark the suite using the file provided, interpreted in raw mode.
+///
+/// Raw mode means that the files lines are parsed line by line, each representing a
+/// message that the CLI sends directly to the Suite.
+/// Implicitly sends OutgoingMessage::Done at the end.
+///
+/// For information about supported messages an their (de)serialization check the common create.
+///
+/// # Arguments
+///
+/// * `tty` - path to the tty used to communicate with the suite
+/// * `input_file` - path to the input file
+pub fn benchmark_raw_file(tty: &OsString, input_file: PathBuf) {
+    let mut suite =
+        SuiteConnection::new(SerialConnection::new(tty).expect("Failed to connect to serial"))
+            .expect("Failed to establish valid connection with suite");
+
+    let input_msg = fs::read_to_string(&input_file).expect("Failed to read input file");
+    for (line_num, line) in input_msg.lines().enumerate() {
+        let msg = parse_raw(line);
+
+        if matches!(msg, OutgoingMessage::Invalid(_) | OutgoingMessage::Done) {
+            panic!(
+                "Input file contains invalid messages on line {}",
+                line_num + 1
+            )
+        }
+
+        suite.send_message(&msg)
+    }
+    suite.send_message(&OutgoingMessage::Done);
+
+    let mut output_msg = String::new();
+    loop {
+        match suite.read_message() {
+            Ok(msg) => {
+                output_msg.push_str(&format!("{msg:#?}\n"));
+
+                if matches!(msg, IncomingMessage::Status(SuiteStatus::Done)) {
+                    break;
+                }
+            }
+            Err(_) => {
```

Comment:
Maybe this should be cleaned up in a better manner if the suite on the hardware runs infinitely? What I mean is if the communication can be restarted and there's pending jobs on hardware they might need to be stopped or discarded. Or do you think this will never happen?

Reply:
Yes it is possible to happen, an example:

1. CLI sends GetStatus as part of establishing the connection
2. Suite replies to (1) with Ready -> Connection established
3. CLI sends GetStatus as part of the benchmarking process
4. CLI sends some Benchmark request
5. CLI sends Done
6. CLI looses connection
7. new CLI sends GetStatus as part of establishing the connection
8. Suite replies to (3) with Ready -> Connection established
9. new CLI sends some other Benchmark request
10. new CLI sends Done
11. Suite replies to (4) with BenchmarkingResults
12. Suite replies to (5) with Done

The new CLI thinks everything worked sucessfully but it recieved the wrong results.
The problem with discarding previous messages on the suite is that due to the communication over the uart, the suite behaves like a fifo.

It should suffice removing the manual status requests for now, since they are similar to "Done" only used for managing the connection.

An alternative would be to add a sender id to every request from the CLI and include this id in every response from the suite. One problem with this is the startup, because during the booting process the opentitan rom takes some time where every message sent to the suite's uart is lost. This is currently circumvented by the suite sending a Ready message once it starts which potentially triggers any CLI that tries to start a connection.

Reply:
I see, makes sense. You could just have TODO for this if you want, incase we see some problems in this case.